src/prepared_statement.c

```diff
@@ -0,0 +1,474 @@
+	ps->query_id = next_unique_query_id;
+	ps->use_count = 0;
+	memcpy(ps->query_with_params, pkt->query, query_len);
+	memcpy(&ps->query_with_params[query_len], pkt->parameter_types_bytes, num_parameters);
+	return ps;
+}
+
+
+void
+ps_init(void)
+{
+	server_ps_cache = slab_create("server_ps_cache", sizeof(PgServerPreparedStatement), 0, NULL, USUAL_ALLOC);
+}
+
+void
+ps_cleanup(void)
+{
+	slab_destroy(server_ps_cache);
+	server_ps_cache = NULL;
+}
+
+
+static PgClientPreparedStatement*
+create_client_prepared_statement(char const* name, PgPreparedStatement* ps)
+{
+	size_t name_len = strlen(name) + 1;
+	PgClientPreparedStatement *client_ps = (PgClientPreparedStatement*)malloc(sizeof(PgClientPreparedStatement) + name_len);
+	memcpy(client_ps->stmt_name, name, name_len);
+	client_ps->ps = ps;
+	ps->use_count += 1;
+	return client_ps;
+}
+
+static PgServerPreparedStatement*
+create_server_prepared_statement(PgPreparedStatement* ps)
+{
+	PgServerPreparedStatement *server_ps = (PgServerPreparedStatement*)slab_alloc(server_ps_cache);
+	server_ps->ps = ps;
+	server_ps->query_id = ps->query_id;
+	ps->use_count += 1;
+	return server_ps;
+}
+
+static PgPreparedStatement*
+get_prepared_statement(PgParsePacket *pkt, bool* found)
+{
+	PgPreparedStatement *ps = create_prepared_statement(pkt);
+	PgPreparedStatement *old_ps = NULL;
+	HASH_FIND(hh, prepared_statements, ps->query_with_params, ps->query_len + ps->param_len, old_ps);
+	if (old_ps != NULL)
+	{
+		free(ps);
+		ps = old_ps;
+		*found = true;
+	}
+	else
+	{
+		HASH_ADD(hh, prepared_statements, query_with_params, ps->query_len + ps->param_len, ps);
+		ps->stmt_name_len = (uint8_t)snprintf(ps->stmt_name, sizeof ps->stmt_name, "PS_%u", ps->query_id);
+		next_unique_query_id += 1;
+		*found = false;
+	}
+	return ps;
+}
+
+static void
+extract_parse_packet(PgPreparedStatement* ps, PgParsePacket *pkt)
+{
+	pkt->query = ps->query_with_params;
+	pkt->num_parameters = ps->param_len/4;
+	pkt->parameter_types_bytes = (uint8_t *)ps->query_with_params + ps->query_len;
+}
+
+static bool
+register_prepared_statement(PgSocket *client, PgSocket *server, PgServerPreparedStatement *server_ps)
+{
+	struct PgServerPreparedStatement *current, *tmp;
+	unsigned int cached_query_count = HASH_COUNT(server->server_prepared_statements);
+
+	if (cached_query_count >= (unsigned int)cf_prepared_statement_cache_size)
+	{
+        int res;
+		/* Remove least recently used statement */
+		HASH_ITER(hh, server->server_prepared_statements, current, tmp) {
+			HASH_DEL(server->server_prepared_statements, current);
+			cached_query_count--;
+			if (--current->ps->use_count == 0)
+			{
+				HASH_DEL(prepared_statements, current->ps);
+				free(current->ps);
+			}
+			slog_noise(server, "prepared statement '%s' deleted from server cache", current->ps->stmt_name);
+			slab_free(server_ps_cache, current);
+
+			// TODO: Consider what happens when this loop fails midway through.
+			QUEUE_CloseStmt(res, client, server, current->ps->stmt_name);
```

Comment:
So what I mean with this is that there's a general design thing to keep in mind with this `sbuf_queue_packet` functionality. If anything fails the packet is not actually sent, as a sort of rollback mechanism. So that the code to handle the packet can be executed again on failure. The tricky part is to make sure all other state is also rolled back when a failure happens, or never happened in the first place. In this case to achieve that it's probably easiest to loop over the hash twice:
1. The first time to create and queue all the CloseStmt packets (because these might allocate and are thus fallible)
2. Remove the prepared statements.

But actually I think the much easier solution is to simply close the client connection in case of OOM. I think I prefer that, for its simplicity to reason about and ease of implementation. I actually think a retry won't even occur in most cases, because it's only retried when new data is received on the read socket.

Reply:
I actually think probably all failures inside any of the handle_xxx_command functions could be considered fatal errors. We should check, but I don't think a retry would solve any of them. If that's indeed the case we can simply call `disconnect_client` at their callsite on failures.